File: pyclient/mmux/kv/replica_chain_client.py
from mmux.kv import block_request_service
from mmux.kv.block_client import BlockClient
from mmux.kv.compat import b, bytes_to_str
from mmux.kv.kv_ops import KVOpType, op_type, KVOps


class ReplicaChainClient:
    class LockedClient:
        def __init__(self, parent):
            self.parent = parent
            response = b(self.run_command(KVOps.lock, [])[0])
            if response != b("!ok"):
                self.redirecting = True
                self.redirect_chain = [bytes_to_str(x) for x in response[1:].split(b('!'))]
            else:
                self.redirecting = False
                self.redirect_chain = []

        def __del__(self):
            self.unlock()

        def unlock(self):
            self.run_command(KVOps.unlock, [])

        def get_chain(self):
            return self.parent.get_chain()

        def is_redirecting(self):
            return self.redirecting

        def get_redirect_chain(self):
            return self.redirect_chain

        def send_cmd(self, cmd_id, args):
            self.parent.send_cmd(cmd_id, args)

        def recv_cmd(self):
            return self.parent.recv_cmd()

        def run_command(self, cmd_id, args):
            return self.parent.run_command(cmd_id, args)

        def run_command_redirected(self, cmd_id, args):
            return self.parent.run_command_redirected(cmd_id, args)

    def __init__(self, client_cache, chain, request_timeout_s=3.0):
        self.seq = block_request_service.sequence_id(-1, 0, -1)
        self.chain = chain
        self.request_timeout_s = request_timeout_s
        h_host, h_port, _, _, _, h_bid = chain[0].split(':')
        self.head = BlockClient(client_cache, h_host, int(h_port), int(h_bid))
        self.seq.client_id = self.head.get_client_id()
        if len(chain) == 1:
            self.tail = self.head
        else:
            t_host, t_port, _, _, _, t_bid = chain[-1].split(':')
            self.tail = BlockClient(client_cache, t_host, int(t_port), int(t_bid))
        self.response_reader = self.tail.get_response_reader(self.seq.client_id)
        self.response_cache = {}
        self.in_flight = False

    def get_chain(self):
        return self.chain

    def _send_cmd(self, client, cmd_id, args):
        if self.in_flight:
            raise RuntimeError("Cannot have more than one request in-flight")
        client.send_request(self.seq, cmd_id, args)
        self.in_flight = True

    def send_cmd(self, cmd_id, args):
        if op_type(cmd_id) == KVOpType.accessor:
            self._send_cmd(self.tail, cmd_id, args)
        else:
            self._send_cmd(self.head, cmd_id, args)

    def _recv_cmd(self):
        rseq, result = self.response_reader.recv_response()
        if self.seq.client_seq_no != rseq:
            raise RuntimeError("SEQ: Expected={} Received={}".format(self.seq.client_seq_no, rseq))
        self.seq.client_seq_no += 1
        self.in_flight = False
        return result

    def recv_cmd(self):
        return self._recv_cmd()

    def _run_command(self, client, cmd_id, args):
        self._send_cmd(client, cmd_id, args)
        return self._recv_cmd()

    def run_command(self, cmd_id, args):
        if op_type(cmd_id) == KVOpType.accessor:
            return self._run_command(self.tail, cmd_id, args)
        else:
            return self._run_command(self.head, cmd_id, args)

    def _run_command_redirected(self, client, cmd_id, args):
        args.append("!redirected")
        self._send_cmd(client, cmd_id, args)
        return self._recv_cmd()

    def run_command_redirected(self, cmd_id, args):
        if op_type(cmd_id) == KVOpType.accessor:
            return self._run_command_redirected(self.tail, cmd_id, args)
        else:
            return self._run_command_redirected(self.head, cmd_id, args)

File: javaclient/core/src/main/java/mmux/kv/ReplicaChainClient.java
package mmux.kv;

import java.io.Closeable;
import java.nio.ByteBuffer;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;
import mmux.kv.BlockClient.CommandResponse;
import mmux.kv.BlockClient.CommandResponseReader;
import mmux.kv.BlockNameParser.BlockMetadata;
import org.apache.thrift.TException;

public class ReplicaChainClient implements Closeable {

  public class LockedClient implements Closeable {

    private ReplicaChainClient parent;
    private boolean redirecting;
    private List<String> redirectChain;

    LockedClient(ReplicaChainClient parent) throws TException {
      this.parent = parent;
      String response = StandardCharsets.UTF_8
          .decode(runCommand(KVOps.LOCK, Collections.emptyList()).get(0)).toString();
      if (!response.equals("!ok")) {
        this.redirecting = true;
        String[] parts = response.split("!");
        this.redirectChain = new ArrayList<>(parts.length - 1);
        this.redirectChain.addAll(Arrays.asList(parts).subList(2, parts.length));
      } else {
        this.redirecting = false;
        this.redirectChain = null;
      }
    }

    void unlock() throws TException {
      runCommand(KVOps.UNLOCK, Collections.emptyList());
    }

    void sendCommandRequest(int cmdId, List<ByteBuffer> args) throws TException {
      parent.sendCommandRequest(cmdId, args);
    }

    List<ByteBuffer> receiveCommandResponse() throws TException {
      return parent.receiveCommandResponse();
    }

    List<ByteBuffer> runCommand(int cmdId, List<ByteBuffer> args) throws TException {
      return parent.runCommand(cmdId, args);
    }

    List<ByteBuffer> runCommandRedirected(int cmdId, List<ByteBuffer> args) throws TException {
      return parent.runCommandRedirected(cmdId, args);
    }

    public boolean isRedirecting() {
      return redirecting;
    }

    public List<String> getRedirectChain() {
      return redirectChain;
    }

    public List<String> getChain() {
      return parent.getChain();
    }

    @Override
    public void close() {
      try {
        unlock();
      } catch (TException ignored) {
      }
    }
  }

  private sequence_id seq;
  private List<String> chain;
  private BlockClient head;
  private BlockClient tail;
  private CommandResponseReader responseReader;
  private BlockClientCache cache;
  private boolean inFlight;

  ReplicaChainClient(BlockClientCache cache, List<String> chain) throws TException {
    if (chain == null || chain.size() == 0) {
      throw new IllegalArgumentException("Chain length must be >= 1");
    }
    this.cache = cache;
    this.chain = chain;
    this.inFlight = false;
    this.seq = new sequence_id(-1, 0, -1);
    connect();
  }

  @Override
  public void close() {
    head.close();
    tail.close();
  }

  public List<String> getChain() {
    return chain;
  }

  public LockedClient lock() throws TException {
    return new LockedClient(this);
  }

  private void sendCommandRequest(BlockClient client, int cmdId, List<ByteBuffer> args)
      throws TException {
    if (inFlight) {
      throw new IllegalStateException("Cannot have more than one request in-flight");
    }
    client.sendCommandRequest(seq, cmdId, args);
    inFlight = true;
  }

  void sendCommandRequest(int cmdId, List<ByteBuffer> args) throws TException {
    if (KVOpType.opType(cmdId) == KVOpType.accessor) {
      sendCommandRequest(tail, cmdId, args);
    } else {
      sendCommandRequest(head, cmdId, args);
    }
  }

  List<ByteBuffer> receiveCommandResponse() throws TException {
    CommandResponse response = responseReader.recieveResponse();

    if (response.clientSeqNo != seq.getClientSeqNo()) {
      throw new IllegalStateException(
          "SEQ: Expected=" + seq.getClientSeqNo() + " Received=" + response.clientSeqNo);
    }
    seq.client_seq_no++;
    inFlight = false;
    return response.result;
  }

  List<ByteBuffer> runCommand(BlockClient client, int cmdId, List<ByteBuffer> args)
      throws TException {
    sendCommandRequest(client, cmdId, args);
    return receiveCommandResponse();
  }

  List<ByteBuffer> runCommand(int cmdId, List<ByteBuffer> args) throws TException {
    if (KVOpType.opType(cmdId) == KVOpType.accessor) {
      return runCommand(tail, cmdId, args);
    } else {
      return runCommand(head, cmdId, args);
    }
  }

  private List<ByteBuffer> runCommandRedirected(BlockClient client, int cmdId,
      List<ByteBuffer> args)
      throws TException {
    List<ByteBuffer> newArgs = new ArrayList<>(args);
    newArgs.add(ByteBuffer.wrap("!redirected".getBytes()));
    return runCommand(client, cmdId, newArgs);
  }

  List<ByteBuffer> runCommandRedirected(int cmdId, List<ByteBuffer> args) throws TException {
    if (KVOpType.opType(cmdId) == KVOpType.accessor) {
      return runCommandRedirected(tail, cmdId, args);
    } else {
      return runCommandRedirected(head, cmdId, args);
    }
  }

  BlockClient getHead() {
    return head;
  }

  BlockClient getTail() {
    return tail;
  }

  private void connect() throws TException {
    BlockMetadata h = BlockNameParser.parse(chain.get(0));
    this.head = new BlockClient(cache, h.getHost(), h.getServicePort(), h.getBlockId());
    this.seq.setClientId(this.head.getClientId());
    if (chain.size() == 1) {
      this.tail = this.head;
    } else {
      BlockMetadata t = BlockNameParser.parse(chain.get(chain.size() - 1));
      this.tail = new BlockClient(cache, t.getHost(), t.getServicePort(), t.getBlockId());
    }
    this.responseReader = this.tail.newCommandResponseReader(seq.getClientId());
  }
}
